src/OrchardCore/OrchardCore/Extensions/ExtensionManager.cs

```diff
@@ -298,11 +298,11 @@ private async Task EnsureInitializedAsync()
                 var loadedExtensions = new ConcurrentDictionary<string, ExtensionEntry>();
 
                 // Load all extensions in parallel
-                await modules.ForEachAsync((module) =>
+                Parallel.ForEach(modules, (module, cancellationToken) =>
```

Comment:
Why not use `Parallel.ForEachAsync()` here too?

Reply:
the delegate is synchronous. No need for asynchronous

Reply:
`EnsureInitializedAsync()` is though so it can make use of `async`.

Reply:
but that is done before the parallel call. The parallel call does not need async

Reply:
`Parallel.ForEach()` blocks the calling thread, since it's synchronous. So, that thread will just wait, like synchronous IO does. `Parallel.ForEachAsync()` being async, the calling thread won't be blocked (and can thus be used to e.g. serve as one of the threads used for the loop).

Reply:
>  the calling thread won't be blocked (and can thus be used to e.g. serve as one of the threads used for the loop)

But it won't because there is no async code in the parallel lambdas.

Reply:
What will the thread of `EnsureInitializedAsync()` do until `Parallel.ForEach()` returns?

Reply:
The only other thing we could change is that since nothing is asynchronous we can remove the async pattern on this method. And replace the semaphore with a lock. To remove any assumption that this is actually async.